### PR TITLE
Support Iceberg row-level delete and update

### PR DESCRIPTION
**Pull request by jackye1995**
This PR continues the effort of #8534 and #8565 to provide full support for reading Iceberg position and equality deletes, and writing Iceberg position deletes in Parquet.

I have added some tests to ensure the correctness of the implementation, and I will continue to add more tests in the following days. I will leave some comments in the code as discussion points.

This is a big PR and we can separate it to multiple for actual contribution, but anyone interested can also try this patch out, I have made sure all related tests pass.

### CODE REVIEW

**Comment by jackye1995** [comment]
@alexjo2144 @findepi @losipiuk @phd3

**Comment by jackye1995** [comment]
@electrum @rdblue

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java [hunk 1]
```diff
@@ -65,6 +72,11 @@ public IcebergTableProperties(IcebergConfig icebergConfig)
                         "File system location URI for the table",
                         null,
                         false))
+                .add(integerProperty(
```
Comment (jackye1995) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:75]: Added support for users to directly create a v2 table using property `format_version`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/IcebergPositionDeletePageSink.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    private final String outputPath;
+    private final JsonCodec<CommitTaskData> jsonCodec;
+    private final IcebergFileWriter writer;
+
+    private long writtenBytes;
+    private long systemMemoryUsage;
+    private long validationCpuNanos;
+    private boolean writtenData;
+
+    public IcebergPositionDeletePageSink(
```
Comment (jackye1995) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/IcebergPositionDeletePageSink.java:64]: My understanding is that all deletes received will always come from the same split, which means the same partition, so there is no need to use the fanout writer. Please let me know if it is not the case.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/IcebergPositionDeletePageSink.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
+        Collection<Slice> commitTasks = new ArrayList<>();
+        if (writtenData) {
+            writer.commit();
+            CommitTaskData task = new CommitTaskData(
+                    outputPath,
+                    writer.getWrittenBytes(),
+                    new MetricsWrapper(writer.getMetrics()),
```
Comment (jackye1995) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/delete/IcebergPositionDeletePageSink.java:118]: Ideally columns of position delete should use full metrics mode. I will add that feature later.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import org.apache.iceberg.FileScanTask;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.util.SerializationUtil.deserializeFromBase64;
+import static org.apache.iceberg.util.SerializationUtil.serializeToBase64;
+
+public class IcebergFileScanTaskWrapper
```
Comment (jackye1995) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java:24]: As we discussed, Iceberg is adding full Jackson support for its object models. This wrapper will be updated to use Jackson serialization once that is completed. Meanwhile I delegate everything to Java serialization that Iceberg provides. Because of that, some related classes are added `Serializable` or `Externalizable` implementations.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 5]
```diff
@@ -622,7 +745,13 @@ private ConnectorTableMetadata getTableMetadata(ConnectorSession session, Schema
     @Override
     public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle)
```
Comment (jackye1995) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:746]: By adding row-level deletes, now some behaviors of `DELETE` becomes unclear to users. Some queries run as metadata delete, some as row-level delete. We might want to introduce the following configs:
1. if metadata delete should be used when possible
2. for metadata delete, should it use `DeleteFiles` API or global equality delete
3. for row-level delete, should it use position delete or equality delete

We know Trino architecture fits position delete better than equality delete, but please read document https://github.com/apache/iceberg/pull/3432 for the tradeoff between the 2. There might be use cases that people desire to use equality delete instead.

We will probably add these configs when the feature request comes.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java (hunk 1 above)
Comment (HiLany) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:75]: Does this mean that I can define the iceberg table spec in the **with** clause by specifying **format_version**?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 5 above)
Comment (lhofhansl) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:746]: I think it's better to avoid new setting if possible. So questions:

1. If metadata delete is possible you'd always want that, right?
2. If so, why would anybody want the equality delete when DeleteFiles works?
3. Not clear here. That's seems to be the only one we want to configure. And probably on a query-by-query basis.

Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:746]: 1. yes, we should extend the feature of metadata delete to also consider Iceberg hidden partitions. As of today it only works for direct partition columns, and deleting based on partition transform predicate would result in row-level delete. That is the "unclear behavior to user" I meant.
2. because it does not touch data file, so will not cause commit conflict.
3. agree, I think we will need session config and ideally per-query config for this (if there is a need from community). But for now position delete definitely works the best with Trino row-level delete design.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java [hunk 6]
```diff
@@ -25,11 +25,26 @@
...
         implements ColumnHandle
 {
+    public static final IcebergColumnHandle ROW_POSITION_HANDLE = new IcebergColumnHandle(
+            primitiveColumnIdentity(ROW_POSITION.fieldId(), ROW_POSITION.name()),
+            BIGINT,
+            ImmutableList.of(),
+            BIGINT,
+            Optional.empty());
+
+    // use Integer.MIN_VALUE as $row_id field ID, which is currently not reserved by Iceberg
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:44]: would we know if iceberg starts using this id internally? Would sth break?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java [hunk 7]
```diff
@@ -29,6 +30,9 @@
 
 public class IcebergConfig
 {
+    public static final int FORMAT_VERSION_SUPPORT_MIN = 1;
+    public static final int FORMAT_VERSION_SUPPORT_MAX = 2;
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:34]: MIN/MAX_SUPPORTED_VERSION

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 8]
```diff
@@ -729,6 +957,47 @@ public ProjectedLayout getFieldLayout(OrcColumn orcColumn)
...
+                outputColumnMapping.add(projectedColumnCount);
+                projectedIsRowPositionChannel.add(column.isIcebergRowPositionMetadataColumn());
+                projectedColumnCount++;
+            }
+            else {
+                outputColumnMapping.add(mapped);
+            }
+        }
+
+        Optional<ReaderColumns> readerColumns = Optional.of(new ReaderColumns(projectedColumns.build(), outputColumnMapping.build()));
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:997]: why optional if never empty?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 9]
```diff
@@ -169,22 +211,117 @@ public ConnectorPageSource createPageSource(
...
+        List<IcebergColumnHandle> fileReadColumns = new ArrayList<>();
+        List<Boolean> isRowPositionChannel = new ArrayList<>();
+
+        // 1. non-partition queried columns
+        Block[] queriedColumnPrefillValues = new Block[queriedColumns.size()];
+        int[] queriedColumnFileReadChannels = new int[queriedColumns.size()];
+        boolean isDeleteOrUpdateQuery = false;
+        for (int idx = 0; idx < queriedColumns.size(); idx++) {
+            IcebergColumnHandle column = queriedColumns.get(idx);
+            if (column.isTrinoRowIdColumn()) {
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:240]: Should we be more explicit and encode fact that we are doing DELETE/UPDATE in `IcebergTableHandle` instead of inferring it from list of columns?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 10]
```diff
@@ -169,22 +211,117 @@ public ConnectorPageSource createPageSource(
...
-        List<IcebergColumnHandle> icebergColumns = columns.stream()
+        Table icebergTable = table.getTable();
+        FileScanTask task = split.getTask();
+        List<IcebergColumnHandle> updateColumns = table.getUpdateColumns();
+        List<IcebergColumnHandle> queriedColumns = columns.stream()
                 .map(IcebergColumnHandle.class::cast)
                 .collect(toImmutableList());
 
         Map<Integer, Optional<String>> partitionKeys = split.getPartitionKeys();
+        Optional<StructLike> partition = task.spec().isUnpartitioned() ? Optional.empty() : Optional.of(task.file().partition());
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:222]: The whole `createPageSource` is hard to follow. Block of code with lots of raw index operations. Would be great to add some higher level structure to it.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 11]
```diff
@@ -193,16 +330,82 @@ public ConnectorPageSource createPageSource(
...
         Optional<ReaderProjectionsAdapter> projectionsAdapter = dataPageSource.getReaderColumns().map(readerColumns ->
                 new ReaderProjectionsAdapter(
-                        regularColumns,
+                        fileReadColumns,
                         readerColumns,
                         column -> ((IcebergColumnHandle) column).getType(),
                         IcebergPageSourceProvider::applyProjection));
-        return new IcebergPageSource(icebergColumns, partitionKeys, dataPageSource.get(), projectionsAdapter);
+
+        return new IcebergPageSource(
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:369]: nit: wrap after each arg

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 12]
```diff
@@ -193,16 +330,82 @@ public ConnectorPageSource createPageSource(
...
+                allTableColumnChannels, posDeleteSink, updateRowSink, projectionsAdapter,
+                isDeleteOrUpdateQuery, isUpdateQuery);
+    }
+
+    private void prefillPartitionValuesAndCompleteFileReadChannels(
+            int idx,
+            IcebergColumnHandle column,
+            List<Integer> fileReadColumnIds,
+            List<IcebergColumnHandle> fileReadColumns,
+            Object[] prefillValues,
```
Comment (losipiuk) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:383]: Block[] ?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java (hunk 4 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java:24]: What is the timeline for that. Will it be feasible to wait for Jackson serialization before merging this PR?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java (hunk 1 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java:75]: yes

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java (hunk 4 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java:24]: yeah it's up to the community, I am just posting this out for review first while working on that. The 0.13 train is going to be out soon so we will target 0.14 for this support, so likely around Jan 2022.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java (hunk 6 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:44]: Because Iceberg column name and type could change, comparing column ID is needed for equality check. Using a duplicated column ID would cause issue in that area.

All the Iceberg metadata columns have IDs counting down from `Integer.MAX_VALUE`. That's why I am choosing `Integer.MIN_VALUE` here. If there is any risk of conflicting IDs, Iceberg will inform the Trino community, and switching to another ID is fully backwards compatible.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java (hunk 10 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:222]: Yes I agree, I can explore that a bit more to see what's the best way to refactor this to be more readable.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 13]
```diff
@@ -98,6 +134,38 @@ public boolean isFinished()
         return delegate.isFinished();
     }
 
+    private Block[] buildBlocksFromPage(Page page, Block[] prefillValues, int[] fileReadChannels)
+    {
+        int batchSize = page.getPositionCount();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:139]: batchSize -> positionCount

but actually better inline

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 14]
```diff
@@ -98,6 +134,38 @@ public boolean isFinished()
...
+            else {
+                blocks[i] = page.getBlock(fileReadChannel);
+            }
+        }
+        return blocks;
+    }
+
+    private Block buildRowIdBlock(Page page)
+    {
+        int batchSize = page.getPositionCount();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:158]: batchSize -> positionCount

but actually better inline

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 15]
```diff
@@ -98,6 +134,38 @@ public boolean isFinished()
...
+    private Block[] buildBlocksFromPage(Page page, Block[] prefillValues, int[] fileReadChannels)
+    {
+        int batchSize = page.getPositionCount();
+        Block[] blocks = new Block[prefillValues.length];
+        for (int i = 0; i < prefillValues.length; i++) {
+            int fileReadChannel = fileReadChannels[i];
+            if (fileReadChannel == -1) {
+                blocks[i] = new RunLengthEncodedBlock(prefillValues[i], batchSize);
+            }
+            else if (fileReadChannel == -2) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:146]: what do these -1, -2 mean? document. define constants.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 16]
```diff
@@ -98,6 +134,38 @@ public boolean isFinished()
...
+                blocks[i] = buildRowIdBlock(page);
+            }
+            else {
+                blocks[i] = page.getBlock(fileReadChannel);
+            }
+        }
+        return blocks;
+    }
+
+    private Block buildRowIdBlock(Page page)
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:156]: move above `buildBlocksFromPage`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 17]
```diff
@@ -98,6 +134,38 @@ public boolean isFinished()
...
+    {
+        int batchSize = page.getPositionCount();
+        Block[] blocks = new Block[nonUpdateColumnPrefillValues.length + 2];
+        blocks[0] = new RunLengthEncodedBlock(filePathBlock, batchSize);
+        blocks[1] = page.getBlock(rowPositionChannel);
+        Block[] nonUpdateColumnBlocks = buildBlocksFromPage(page, nonUpdateColumnPrefillValues, nonUpdateColumnFileReadChannels);
+        for (int i = 0; i < nonUpdateColumnBlocks.length; i++) {
+            blocks[i + 2] = nonUpdateColumnBlocks[i];
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:164]: `System.arraycopy`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 18]
```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
...
+    public void deleteRows(Block rowIds)
+    {
+        RowBlock rows = resolveRowIdBlock(rowIds);
+        posDeleteSink.appendPage(new Page(rows.getPositionCount(), rows.getChildren().toArray(new Block[0])));
+    }
+
+    @Override
+    public void updateRows(Page page, List<Integer> columnValueAndRowIdChannels)
+    {
+        int batchSize = page.getPositionCount();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:213]: `positionCount`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 19]
```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
...
+        Block rowIdRawBlock = page.getBlock(columnValueAndRowIdChannels.get(columnValueAndRowIdChannels.size() - 1));
+        RowBlock rowIdBlock = resolveRowIdBlock(rowIdRawBlock);
+        Block[] rowIdFieldBlocks = rowIdBlock.getChildren().toArray(new Block[0]);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:216]: `Block.getChildren()` is probably not correct. 
see https://github.com/trinodb/trino/pull/9354
you might want to check with @djsstarburst , but it seems ColumnarRow should be used here.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 20]
```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
...
+    private RowBlock resolveRowIdBlock(Block rowIds)
+    {
+        RowBlock rows;
+        if (rowIds instanceof RowBlock) {
+            rows = (RowBlock) rowIds;
+        }
+        else if (rowIds instanceof DictionaryBlock) {
+            Block dictionary = ((DictionaryBlock) rowIds).getDictionary();
+            if (dictionary instanceof RowBlock) {
+                rows = (RowBlock) dictionary;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:246]: The semantics of `resolveRowIdBlock` are ill-defined. You get "some block, with some values".
the underlying `DictionaryBlock.getDictionary` may contain bogus data, or pretty much anything

remove this method.

you *probably* need it because you didn't the proper class, see the other comment about `Block.getChildren`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 21]
```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
...
+        }
+        return rows;
+    }
+
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
+        try {
+            Collection<Slice> slices = new ArrayList<>();
+            if (posDeleteSink != null) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:263]: if the field is nullable, annotate the field and constructor param with `@Nullable`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 22]
```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
...
+        catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void abort()
+    {
+        posDeleteSink.abort();
+        updateRowSink.abort();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:280]: `finish()` assumes the fields are nullable.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java [hunk 23]
```diff
@@ -128,6 +200,87 @@ public Page getNextPage()
...
+        close();
+    }
+
     @Override
     public void close()
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:285]: close posDeleteSink, updateRowSink too

**Review by findepi — COMMENTED** [review]
(skimming)

**Comment by littleWT** [comment]
> This PR continues the effort of #8534 and #8565 to provide full support for reading Iceberg position and equality deletes, and writing Iceberg position deletes in Parquet.
> 
> I have added some tests to ensure the correctness of the implementation, and I will continue to add more tests in the following days. I will leave some comments in the code as discussion points.
> 
> This is a big PR and we can separate it to multiple for actual contribution, but anyone interested can also try this patch out, I have made sure all related tests pass.

HI，Recently，I tested this PR on flink CDC data  in iceberg，and found that  when no update/delete data it would be ok and query very fast。But it can not work when there are  a lot of  update/delete data.

**Comment by jackye1995** [comment]
thanks for the information @littleWT !

This PR focuses on getting the feature in, I would expect the performance to be not super good because vectorization of reading deletes is still work in progress in Iceberg. Position delete vectorized read will be out in 0.13, equality delete vectorized read will be out in 0.14.

It would be great if you can provide some information related to your test scale for other people to refer.

**Comment by littleWT** [comment]
@jackye1995 Haha,we had talk aboult the problem in Dingding!
We tested about 13 millions rows data and when there are only thounds delete/update rows  the query would not work.
Now we have  combined the CDC data  with SPARK（MOR）， that would be better than before. But not so good.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 24]
```diff
@@ -536,7 +552,114 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
...
+        List<Integer> path = new ArrayList<>();
+        icebergFields.add(DELETE_FILE_PATH);
+        trinoFields.add(RowType.field(DELETE_FILE_PATH.name(), VARCHAR));
+        path.add(DELETE_FILE_PATH.fieldId());
+        icebergFields.add(DELETE_FILE_POS);
+        trinoFields.add(RowType.field(DELETE_FILE_POS.name(), BIGINT));
+        path.add(DELETE_FILE_POS.fieldId());
+
+        // for update, include all the non-update columns
+        if (!updatedColumns.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:573]: Looks like it's empty only in the delete case. Might be clearer making the `updatedColumns` list optional. `empty` in the case of deletes, and then verify the list is never `Optional.of(<empty_list>)`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java (hunk 4 above)
Comment (dain) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/serdes/IcebergFileScanTaskWrapper.java:24]: Java serialization is a hard no for Trino.  It is huge ugly mess with lots of security problems.  Instead, for temporary storage you could do something like annotate the iceberg objects with Jackson annotations or use Jackson features to [serialize third party objects](https://web.archive.org/web/20110829212918/http://wiki.fasterxml.com/JacksonMixInAnnotations).  For long term storage, you should write your own serialization objects and copy in and out of them.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 25]
```diff
@@ -536,7 +552,114 @@ public ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connecto
...
+
+        Types.NestedField icebergRowIdField = Types.NestedField.required(TRINO_ROW_ID_COLUMN_ID, TRINO_ROW_ID_COLUMN_NAME, Types.StructType.of(icebergFields));
+        io.trino.spi.type.Type trinoRowIdType = RowType.from(trinoFields);
+        return new IcebergColumnHandle(createColumnIdentity(icebergRowIdField), trinoRowIdType, ImmutableList.of(), trinoRowIdType, Optional.empty());
+    }
+
+    @Override
+    public ConnectorTableHandle beginUpdate(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> updatedColumns)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:596]: This should throw an unsupported exception if the table is v1 right? Same for `beginDelete`

**Comment by findepi** [comment]
Please add a test with OPTIMIZE after some rows modified with DELETE and UPDATE.

**Comment by alexjo2144** [comment]
@jackye1995 @findepi I'm going to take a look at the conflicts and see if I can split some of this out into smaller chunks. It looks like some of the serialization work for the REST catalog should be useful here too.

**Comment by findepi** [comment]
Thanks @alexjo2144 !
also cc @kbendick re serialization

**Comment by electrum** [comment]
@jackye1995 thanks for your work here. @alexjo2144 can this be closed now, or is there additional code to be merged?

**Comment by alexjo2144** [comment]
So far I've only pulled out the read support pieces from here, I'm working on the merge conflicts for write support now. 

I have a copy of Jack's code though, so I think we can close this PR.

**Comment by findepi** [comment]
Superseded by https://github.com/trinodb/trino/pull/11886
